lib/posix-vfs/vfs.c

```diff
@@ -0,0 +1,2025 @@
+	target = vfs_lookupat(vfs_atroot(NULL, targetpath, pathlen),
+			      targetpath, pathlen, 0);
+	if (unlikely(PTRISERR(target)))
+		return PTR2ERR(target);
+
+	uk_mutex_lock(&mtab.lock);
+	/* Find target in mtab */
+	for (mntidx = 0; mntidx < mtab.top; mntidx++) {
+		if (mtab.tab[mntidx].parent_root == target)
+			inval = 1;
+		if (mtab.tab[mntidx].root == target)
+			break;
+	}
+	UK_ASSERT(!inval);
+	if (mntidx != mtab.top) {
+		const struct uk_file *prev = NULL;
+
+		for (size_t i = mntidx + 1; i < mtab.top; i++)
+			if (unlikely(mtab.tab[i].parent_root == target)) {
+				ret = -EBUSY;
+				goto out;
+			}
+		/* unmount & release */
+		ret = uk_fs_mountat(mtab.tab[mntidx].point, &prev);
+		if (unlikely(ret))
+			goto out;
+		UK_ASSERT(prev == target);
+		uk_file_release(prev);
+		/* rm from mtab */
+		for (size_t i = mntidx + 1; i < mtab.top; i++)
+			mtab.tab[i - 1] = mtab.tab[i];
+		mtab.top--;
+		ret = 0;
+	} else {
+		ret = -EINVAL;
+	}
+out:
+	uk_mutex_unlock(&mtab.lock);
+	uk_file_release(target);
+	return ret;
+}
+
+/* Open */
+
+/**
+ * Lookup a path, following all symlinks; if not exist but could be created,
+ * return a reference to a parent dir, the desired filename, and an optional
+ * reference to the symlink that ensures the filename's lifetime.
+ *
+ * The reference to `atroot` is consumed by this call.
+ */
+static
+int vfs_lookupat_parent(const struct uk_file *atroot,
+			const char *path, size_t pathlen,
+			const struct uk_file **ref, const char **name,
+			size_t *len, const struct uk_file **symref)
+{
+	const struct uk_file *sym = NULL;
+	const struct uk_file *f = atroot;
+	size_t symlvl = 0;
+	int ret;
+
+	do {
+		size_t ppos = pathlen;
+		const struct uk_file *dir;
+		const struct uk_file *target;
+
+		while (ppos && path[ppos - 1] != '/')
+			ppos--;
+		dir = _vfs_lookupat(f, path, ppos, 0,
+#if CONFIG_LIBPOSIX_VFS_SYMLIMIT_DEPTH
+				    symlvl
+#elif CONFIG_LIBPOSIX_VFS_SYMLIMIT_NUM
+				    &symlvl
+#endif /* CONFIG_LIBPOSIX_VFS_SYMLIMIT_NUM */
+		);
+		if (unlikely(PTRISERR(dir))) {
+			ret = PTR2ERR(dir);
+			break;
+		}
+		target = _vfs_lookupat(dir, &path[ppos], pathlen - ppos,
+				       VFS_LOOKUP_NOFOLLOW,
+#if CONFIG_LIBPOSIX_VFS_SYMLIMIT_DEPTH
+				     symlvl
+#elif CONFIG_LIBPOSIX_VFS_SYMLIMIT_NUM
+				     &symlvl
+#endif /* CONFIG_LIBPOSIX_VFS_SYMLIMIT_NUM */
+		);
+		if (PTRISERR(target)) {
+			if (PTR2ERR(target) == -ENOENT) {
+				/* Reached end & not present, ret parent */
+				*ref = dir;
+				*name = &path[ppos];
+				*len = pathlen - ppos;
+				*symref = sym;
+				ret = 1;
+			} else {
+				uk_file_release(dir);
+				ret = PTR2ERR(target);
+				UK_ASSERT(ret < 0);
+			}
+			break;
+		}
+		/* Present; if symlink follow, if not ret success */
+		if (sym)
+			uk_file_release(sym);
+		if (UKFS_FILE_ISSYM(target)) {
+			struct uk_fs_path p = uk_fs_readlink(target);
+
+			uk_file_release(f);
+			f = dir;
+			path = p.s;
+			pathlen = p.len;
+			sym = target;
+			symlvl++;
```

Comment:
Should this maybe also be guarded by `CONFIG_LIBPOSIX_VFS_SYMLIMIT_DEPTH` and be declared with a `__maybe_unused`?